spi/src/main/java/org/opensearch/jobscheduler/spi/LockModel.java

```diff
@@ -88,37 +117,128 @@ public LockModel(final LockModel copyLock, final boolean released) {
+     * @param resourceType          type of job acquiring the lock
+     * @param resource              resource to acquire
+     * @param lockTime              time at which lock is acquired
+     * @param lockDurationSeconds   duration for which lock is valid
+     * @param released              boolean flag for whether lock is released
+     * @param seqNo                 sequence number from Opensearch document
+     * @param primaryTerm           primary term from Opensearch document
+     * @throws IOException          throws IOException if resource fails to serialize
+     */
+    public LockModel(String jobIndexName, String jobId, String resourceType, Map<String, Object> resource, Instant lockTime,
+                     long lockDurationSeconds, boolean released, long seqNo, long primaryTerm) throws IOException {
+        if (resourceType != null) {
+            this.lockId = LockModel.generateResourceLockId(jobIndexName, resourceType, resource);
+            this.jobIndexName = jobIndexName;
+            this.jobId = jobId;
+            this.resourceType = resourceType;
+            this.resource = resource;
+            this.lockTime = lockTime;
+            this.lockDurationSeconds = lockDurationSeconds;
+            this.released = released;
+            this.seqNo = seqNo;
+            this.primaryTerm = primaryTerm;
+        }
+        else {
+            this.lockId = jobIndexName + LOCK_ID_DELIMITER + jobId;
+            this.jobIndexName = jobIndexName;
+            this.jobId = jobId;
+            this.lockTime = lockTime;
+            this.lockDurationSeconds = lockDurationSeconds;
+            this.released = released;
+            this.seqNo = seqNo;
+            this.primaryTerm = primaryTerm;
+            this.resource = null;
+            this.resourceType = null;
+        }
+
     }
 
     public static String generateLockId(String jobIndexName, String jobId) {
-        return jobIndexName + LOCK_ID_DELIMITR + jobId;
+        return jobIndexName + LOCK_ID_DELIMITER + jobId;
+    }
+
+    public static String generateResourceLockId(String jobIndexName, String resourceType, Map<String, Object> resource) throws IOException {
+        try {
+            byte[] resourceAsBytes = serialize(resource);
+            MurmurHash3.Hash128 hash = MurmurHash3.hash128(resourceAsBytes, 0, resourceAsBytes.length, 0,
+                    new MurmurHash3.Hash128());
+            byte[] resourceHashBytes = ByteBuffer.allocate(16).putLong(hash.h1).putLong(hash.h2).array();
+            String resourceAsHashString = Base64.getUrlEncoder().withoutPadding().encodeToString(resourceHashBytes);
+            return jobIndexName + LOCK_ID_DELIMITER + resourceType + LOCK_ID_DELIMITER + resourceAsHashString;
+        } catch (IOException ioException) {
+            ioException.printStackTrace();
+            throw ioException;
+        }
+    }
+    public static byte[] serialize(Object obj) throws IOException {
```

Comment:
format: missing empty line before

Also can you make this a private method? see this is only being used by `generateResourceLockId` Not really sure if this needs to be public